File: 475-heaters/475-heaters.cs
public class Solution {
    public int FindRadius(int[] houses, int[] heaters) {
            Array.Sort(houses);
            Array.Sort(heaters);

            var ls = new List<int>();
            for (int i = 0; i < houses.Length; i++)
            {
                ls.Add(int.MaxValue);
            }

            //dist to nearest rhs heater
            for (int i = 0, h = 0; i < houses.Length && h < heaters.Length;)
            {
                if (houses[i] <= heaters[h])
                {
                    ls[i] = heaters[h] - houses[i];
                    i++;
                }
                else
                {
                    h++;
                }
            }

            //dist to nearest lhs heater
            for (int i = houses.Length - 1, h = heaters.Length - 1; i >= 0 && h >= 0;)
            {
                if (houses[i] >= heaters[h])
                {
                    ls[i] = Math.Min(ls[i], houses[i] - heaters[h]);
                    i--;
                }
                else
                {
                    h--;
                }
            }

            return ls.Max();
    }
}

File: 475-heaters/475-heaters.cpp
class Solution {
public:
    int findRadius(vector<int>& houses, vector<int>& heaters) {
        
        int l = houses.size();
        sort(houses.begin(),houses.end());
        sort(heaters.begin(),heaters.end());

        vector<int> d(l,1e9+7);

        for (int i = 0, h=0; i < houses.size() && h < heaters.size();)
        {
            if(houses[i]<=heaters[h]){
                d[i] = heaters[h]-houses[i];
                i++;
            }else{
                h++;
            }
        }

        for (int i = houses.size()-1, h = heaters.size() - 1; i >=0 && h >= 0;)
        {
            if(houses[i]>=heaters[h]){
                d[i]=min(d[i],houses[i]-heaters[h]);
                i--;
            }else{
                h--;
            }
        }

        return *max_element(d.begin(),d.end());
    }
};